reconcile/utils/oc.py

```diff
@@ -668,6 +682,252 @@ def _run_json(self, cmd, allow_not_found=False):
+        if server is not None and len(server) > 0:
+            self.client = self._get_client(server, token)
+            self.api_kind_version = self.get_api_resources()
+        else:
+            init_api_resources = False
+            init_projects = False
+
+        self.object_clients = {}
+        self.init_projects = init_projects
+        if self.init_projects:
+            self.projects = \
+                [p['metadata']['name']
+                 for p
+                 in self.get_all('Project.project.openshift.io')['items']]
+        self.init_api_resources = init_api_resources
+        if self.init_api_resources:
+            self.api_resources = self.api_kind_version.keys()
+        else:
+            self.api_resources = None
+
+    def _get_client(self, server, token):
+        opts = dict(
+            api_key={'authorization': f'Bearer {token}'},
+            host=server,
+            verify_ssl=False,
+            # default timeout seems to be 1+ minutes and it can't
+            # be changed, so do no retries on connection failure
+            retries=0
+        )
+
+        if self.jump_host is not None:
+            # the ports could be parameterized, but at this point
+            # we only have need of 1 tunnel for 1 service
+            self.jump_host.create_ssh_tunnel(8888, 8888)
+            opts['proxy'] = 'http://localhost:8888'
+
+        configuration = kubernetes.client.Configuration()
+
+        # the kubernetes client configuration takes a limited set
+        # of parameters during initialization, but there are a lot
+        # more options that can be set to tweak the behavior of the
+        # client via instance variables.  We define a set of options
+        # above in the format of var_name:value then set them here
+        # in the configuration object with setattr.
+        for k, v in opts.items():
+            setattr(configuration, k, v)
+
+        k8s_client = kubernetes.client.ApiClient(configuration)
+        try:
+            return DynamicClient(k8s_client)
+        except urllib3.exceptions.MaxRetryError as e:
+            raise StatusCodeError(f"[{self.server}]: {e}")
+
+    def _get_obj_client(self, kind, group_version):
+        key = f'{kind}.{group_version}'
+        if key not in self.object_clients:
+            self.object_clients[key] = self.client.resources.get(
+                api_version=group_version, kind=kind)
+        return self.object_clients[key]
+
+    def _parse_kind(self, kind_name):
+        kind_group = kind_name.split('.', 1)
+        kind = kind_group[0]
+        if kind in self.api_kind_version:
+            group_version = self.api_kind_version[kind][0]
+        else:
+            raise StatusCodeError(f'{self.server}: {kind} does not exist')
+
+        # if a kind_group has more than 1 entry than the kind_name is in
+        # the format kind.apigroup.  Find the apigroup/version that matches
+        # the apigroup passed with the kind_name
```

Comment:
what if we don't find it with the algorithm below? will we send a relevant error message?

Reply:
Actually, no we won't.  If we don't find the kind.apigroup then we will return the group_version that is known to the cluster.  For example:

If kind_name = whosit.whatsit
and api_kind_version contains:
{'whosit':["clusterapigroup"]}
then we would return
kind = whosit
group_version = clusterapigroup

This effectively means that the overrided apigroup doesn't exist on the cluster.  Since we are trying to override explicitly, we probably should throw an exception here.

Reply:
Done